src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs

```diff
@@ -1780,16 +1839,91 @@ LocalDefinition emitGenericReceiver(BoundCall call, out CallKind callKind)
                     _builder.EmitLocalStore(tempOpt);
                     _builder.EmitLocalAddress(tempOpt);
 
-                    if (whenNotNullLabel is not null)
+                    if (whenValueTypeLabel is not null)
                     {
-                        _builder.MarkLabel(whenNotNullLabel);
+                        _builder.MarkLabel(whenValueTypeLabel);
                     }
                 }
 
                 return tempOpt;
             }
         }
 
+        private object TryEmitIsValueTypeBranch(TypeSymbol type, SyntaxNode syntax)
+        {
+            if (Binder.GetWellKnownTypeMember(_module.Compilation, WellKnownMember.System_Type__GetTypeFromHandle, _diagnostics, syntax: syntax, isOptional: false) is MethodSymbol getTypeFromHandle &&
+                Binder.GetWellKnownTypeMember(_module.Compilation, WellKnownMember.System_Type__get_IsValueType, _diagnostics, syntax: syntax, isOptional: false) is MethodSymbol getIsValueType)
+            {
+                _builder.EmitOpCode(ILOpCode.Ldtoken);
+                EmitSymbolToken(type, syntax);
+                _builder.EmitOpCode(ILOpCode.Call, stackAdjustment: 0); // argument off, return value on
+                EmitSymbolToken(getTypeFromHandle, syntax, null);
+                _builder.EmitOpCode(ILOpCode.Call, stackAdjustment: 0); // instance off, return value on
+                EmitSymbolToken(getIsValueType, syntax, null);
+                var whenValueTypeLabel = new object();
+                _builder.EmitBranch(ILOpCode.Brtrue, whenValueTypeLabel);
+
+                return whenValueTypeLabel;
+            }
+
+            return null;
+        }
+
+        private void EmitComplexReceiver(BoundComplexReceiver expression, bool used)
+        {
+            Debug.Assert(!used);
```

Comment:
nit: What's the benefit of only asserting? Throwing seems better.
We can still keep the conditional `AdjustStack` below for the day we could reach with here with used==true.

Reply:
> What's the benefit of only asserting? Throwing seems better.

The implementation is doing the right thing. I do not see any point in throwing. The assert will help us identify the test scenario.